src/Push.cpp

```diff
@@ -0,0 +1,199 @@
+   } else {
+      LOG(INFO) << " zero or negative timeout requested, blocking socket";
+   }
+   nn_setsockopt(mSocket,
+                 NN_SOL_SOCKET,
+                 NN_SNDBUF,
+                 &bufferSize,
+                 sizeof(bufferSize));
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect) {
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = std::string("nn_connect");
+   } else {
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = std::string("nn_bind");
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   LOG(INFO) << connectionType
+             << " socket at: " << mProtocolHandler->GetLocation()
+             << " with timeout: " << timeoutInMs;
+}
+/**
+ * Return the location the socket is bound to
+ * @return
+ */
+std::string Push::GetBinding() const {
+   return mProtocolHandler->GetLocation();
+}
+/**
+ * Fires a basic string
+ */
+void Push::Send(const std::string& data, bool dontWait) {
+   NanoMsg msg(data);
+   SendMessage(msg, dontWait);
+}
+/**
+ * Fires a void pointer
+ */
+void Push::Send(void *data, bool dontWait) {
+   //LOG(INFO) << "sending pointer";
+   NanoMsg msg(data);
+   SendMessage(msg, dontWait);
+}
+/**
+ * Fires a packet hash vector
+ */
+void Push::Send(const std::vector<std::pair<void*, unsigned int>>& data,
+                bool dontWait) {
+   NanoMsg msg(data);
+   //LOG(INFO) << "Sending vector with starting pointer: " << &data[0];
+   //Send(const_cast<std::vector<std::pair<void*, unsigned int> >*>(&data));
+   SendMessage(msg, dontWait);
+}
+/**
+ * private function to send a message using zero copy api
+ */
+void Push::SendMessage(NanoMsg& msg, bool dontWait) {
+   //LOG(INFO) << "sending message";
+   auto num_bytes_sent = nn_send(mSocket,
+                                 &msg.buffer,
+                                 NN_MSG,
+                                 dontWait? NN_DONTWAIT : 0);
+   if (num_bytes_sent < 0) {
+      auto error = nn_errno ();
+      if (error == ETIMEDOUT) {
+         throw std::runtime_error("timed out sending message");
+      } else if (error == EFAULT || error == EBADF || error == ENOTSUP) {
+         // really bad issue,
+         // trying to send a nullptr,
+         // bad socket,
+         // or operation not supported
+         LOG(FATAL) << "fatal nanomsg pull error: "
+                    << std::string(nn_strerror(errno));
+      } else if (error == EFSM || error == EAGAIN) {
+         // failed for now, but you can try again
+         throw std::runtime_error("try again");
```

Comment:
Please do not use exception to return substitute for  status codes. Ref: our c++ guideline.  Please change the status handling paradigm in this PR and of course in the coupled Research PR.